### ISSUE TITLE
High memory usage by each VMI pod.

### ISSUE THREAD

**Issue by kunalkushwaha** [kubevirt/kubevirt#2109]
/kind bug

**What happened**:
When a VMI is created, the total Memory consumption per VM is much more than it is allocated for VM.
Typically I saw the pattern as follows.
- `virt-launcher` consume around `1.2 GB` irrespective of memory allocated to VM
- `qemu` process running VM consume more than twice the memory allocated to VM

**What you expected to happen**:
- memory overhead should be reduced. 

**How to reproduce it (as minimally and precisely as possible)**:
- Create a VMI.
- check memory usage of `virt-launcher` & `qemu`
```
# ps aux --sort -rss | awk '{$5=int(100 * $5/1024/1024)/100"GB";}{ print;}' | grep qemu
qemu 12607 3.1 6.3 2.58GB 763720 ? Sl Mar11 38:39 /usr/bin/qemu-system-x86_64 -name guest=default_testvmi-postgres,debug-threads=on -S -object secret,id=masterKey0,format=raw,file=/var/lib/libvirt/qemu/domain-1-default_testvmi-post/master-key.aes -machine pc-q35-3.1,accel=tcg,usb=off,dump-guest-core=off -cpu EPYC,acpi=on,ss=on,hypervisor=on,erms=on,mpx=on,pcommit=on,clwb=on,pku=on,la57=on,3dnowext=on,3dnow=on,vme=off,fma=off,avx=off,f16c=off,rdrand=off,avx2=off,rdseed=off,sha-ni=off,xsavec=off,fxsr_opt=off,misalignsse=off,3dnowprefetch=off,osvw=off,topoext=off -m 977 -realtime mlock=off -smp 1,sockets=1,cores=1,threads=1 -object iothread,id=iothread1 -uuid 17481baa-7286-4dbd-a5cd-8e2371685a8d -no-user-config -nodefaults -chardev socket,id=charmonitor,fd=23,server,nowait -mon chardev=charmonitor,id=monitor,mode=control -rtc base=utc -no-shutdown -boot strict=on -device pcie-root-port,port=0x10,chassis=1,id=pci.1,bus=pcie.0,multifunction=on,addr=0x2 -device pcie-root-port,port=0x11,chassis=2,id=pci.2,bus=pcie.0,addr=0x2.0x1 -device pcie-root-port,port=0x12,chassis=3,id=pci.3,bus=pcie.0,addr=0x2.0x2 -device pcie-root-port,port=0x13,chassis=4,id=pci.4,bus=pcie.0,addr=0x2.0x3 -device virtio-serial-pci,id=virtio-serial0,bus=pci.2,addr=0x0 -drive file=/var/run/kubevirt-ephemeral-disks/container-disk-data/default/testvmi-postgres/disk_containerdisk/disk-image.qcow2,format=qcow2,if=none,id=drive-ua-containerdisk,cache=none -device virtio-blk-pci,scsi=off,bus=pci.3,addr=0x0,drive=drive-ua-containerdisk,id=ua-containerdisk,bootindex=1,write-cache=on -netdev tap,fd=25,id=hostua-default -device virtio-net-pci,host_mtu=1376,netdev=hostua-default,id=ua-default,mac=52:3f:f1:88:75:ef,bus=pci.1,addr=0x0 -chardev socket,id=charserial0,fd=26,server,nowait -device isa-serial,chardev=charserial0,id=serial0 -chardev socket,id=charchannel0,fd=27,server,nowait -device virtserialport,bus=virtio-serial0.0,nr=1,chardev=charchannel0,id=channel0,name=org.qemu.guest_agent.0 -vnc vnc=unix:/var/run/kubevirt-private/5d232553-43c7-11e9-92d5-0800278346b4/virt-vnc -device VGA,id=video0,vgamem_mb=16,bus=pcie.0,addr=0x1 -sandbox on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny -msg timestamp=on
root 12364 0.0 0.3 1.22GB 46512 ? Sl Mar11 0:39 /usr/bin/virt-launcher --qemu-timeout 5m --name testvmi-postgres --uid 5d232553-43c7-11e9-92d5-0800278346b4 --namespace default --kubevirt-share-dir /var/run/kubevirt --ephemeral-disk-dir /var/run/kubevirt-ephemeral-disks --readiness-file /var/run/kubevirt-infra/healthy --grace-period-seconds 45 --hook-sidecars 0 --less-pvc-space-toleration 10 --use-emulation --no-fork true
root 12327 0.0 0.3 0.94GB 43268 ? Ssl Mar11 0:07 /usr/bin/virt-launcher --qemu-timeout 5m --name testvmi-postgres --uid 5d232553-43c7-11e9-92d5-0800278346b4 --namespace default --kubevirt-share-dir /var/run/kubevirt --ephemeral-disk-dir /var/run/kubevirt-ephemeral-disks --readiness-file /var/run/kubevirt-infra/healthy --grace-period-seconds 45 --hook-sidecars 0 --less-pvc-space-toleration 10 --use-emulation

```

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
```
$ virtctl version
Client Version: version.Info{GitVersion:"v0.14.0", GitCommit:"99129b309dae399061d38924744ded0ec0ea72c9", GitTreeState:"clean", BuildDate:"2019-02-04T13:49:57Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"v0.15.0", GitCommit:"432f91f3a2bdb3c2624377ad327270c9e146d845", GitTreeState:"clean", BuildDate:"2019-03-05T09:55:30Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"linux/amd64"}

```
- Kubernetes version (use `kubectl version`):
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-27T17:05:32Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-24T06:43:59Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}

```
- VM or VMI specifications:
```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  labels:
    kubevirt.io/vm: testvmi-postgres
  name: testvmi-postgres
spec:
  terminationGracePeriodSeconds: 30
  domain:
    resources:
      requests:
        memory: 1024M
    devices:
      disks:
      - name: containerdisk
        disk:
          bus: virtio
      interfaces:
      - name: default
        bridge: {} # connect through a bridge
        ports:
        - name: http
          port: 80
        - name: ssh
          port: 22
  volumes:
  - name: containerdisk
    containerDisk:
      image: kunalkushwaha/vm-postgres:latest
  networks:
  - name: default
    pod: {}

```
- Cloud provider or hardware configuration:
VirtualBox VM.
- OS (e.g. from /etc/os-release):
```
Fedora 29 (Server Edition)
```
- Kernel (e.g. `uname -a`):
```
Linux kube-node 4.18.18-300.fc29.x86_64 #1 SMP Mon Nov 12 03:12:14 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
- Others:

**Comment by rmohr** [comment]
Hi @kunalkushwaha can it be that this is only the virtual memory size? We had a similar report here: #1662. virt-launcher was consuming in reality about 3MB of memory but had virtual-memory of 1.7 gb.

**Comment by kunalkushwaha** [comment]
@rmohr Yes you are correct, coulum 6 is rss and its much less.

Here is the problem I am facing.
- Have kube-node with 12 GB RAM
- With single VM of 1 GB in KubeVirt, I am just left with around 2 GB of Memory.
```
[root@kube-node ~]# vmstat -sSM
        11723 M total memory
         1391 M used memory
         4564 M active memory
         4201 M inactive memory
         2429 M free memory
            2 M buffer memory
         7899 M swap cache
            0 M total swap
            0 M used swap
            0 M free swap

[root@kube-node ~]# free -h
              total        used        free      shared  buff/cache   available
Mem:           11Gi       1.5Gi       2.3Gi       2.0Mi       7.6Gi       9.7Gi
Swap:            0B          0B          0B

```


Problem arises if kubelet _Out Of Resource_ Handling is tweaked for VM workload and it gets triggered irrespective a actual usage. As `swap off` is configured for all nodes in kubernetes, This will lead to issues. 

with two VMs
```
[root@kube-node ~]# free -h
              total        used        free      shared  buff/cache   available
Mem:           11Gi       1.9Gi       148Mi       2.0Mi       9.4Gi       9.2Gi
Swap:            0B          0B          0B

[root@kube-node ~]# vmstat -sSM
        11723 M total memory
         1980 M used memory
         5040 M active memory
         5949 M inactive memory
          161 M free memory
            2 M buffer memory
         9578 M swap cache
            0 M total swap
            0 M used swap
            0 M free swap

```

**Comment by cynepco3hahue** [comment]
In the output of free,

   -

   *Free* memory is the amount of memory which is currently not used for
   anything. This number should be small, because memory which is not used is
   simply wasted.
   -

   *Available* memory is the amount of memory which is available for
   allocation to a new process or to existing processes.

So the output looks totally ok.

**Comment by rmohr** [comment]
@kunalkushwaha are you concerned that the VMIs will get evicted or did you see them getting evicted?

Maybe also this resource is helpful: https://kubevirt.io/user-guide/docs/latest/using-virtual-machines/overcommit.html.

**Comment by rmohr** [comment]
> I faced that previously, and increased memory and Disk to fix the issue.

Did that happen with VMIs or did you face this in general?

**Comment by kunalkushwaha** [comment]
> @kunalkushwaha are you concerned that the VMIs will get evicted or did you see them getting evicted?

My running VMI was getting shutdown. I initially extended the memory by looking at `free` output. Later extended the Disk size also. Problem got fixed. 
I tried to reproduce the issue, by just reducing ram size, and unable to reproduce currently.

while looking at this, I also found [one script](https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/memory-available.sh) mentioned in k8s doc[ eviction-policy](https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#eviction-policy). script suggest, there is sufficient memory available on node. [same explained by @cynepco3hahue too in above comment].

I will try to recreate similar environment where I faced the VMI shutdown issue and see if that problem was only of disk or memory too.
